ion-schema/src/model/constraints/valid_values.rs

```diff
@@ -0,0 +1,363 @@
+use crate::internal_traits::{
+    LoaderContext, ReadFromIsl, ValidateInternal, ValidationContext, WriteAsIsl, WriteContext,
+};
+use crate::ion_extension::ElementExtensions;
+use crate::model::constraints::ConstraintName;
+use crate::model::ranges::IonSchemaRange;
+use crate::model::TypeDefinitionBuilder;
+use crate::result::{invalid_schema_error, IonSchemaResult};
+use crate::{IonSchemaElement, IslVersion, ViolationRecorder};
+use ion_rs::{Decimal, Element, SequenceWriter, Timestamp, ValueWriter};
+use ion_rs::{Value as IonValue, Value};
+use std::fmt::Debug;
+use std::ops::{ControlFlow, RangeBounds};
+
+impl ConstraintName for ValidValues {
+    const CONSTRAINT_NAME: &'static str = "valid_values";
+}
+
+/// The `valid_values` constraint (ref. [ISL 1.0], [ISL 2.0]).
+///
+/// [ISL 1.0]: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#valid_values
+/// [ISL 2.0]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#valid_values
+#[derive(Clone, Debug, PartialEq)]
+pub struct ValidValues {
+    // TODO: Use a Set once Timestamp and Decimal implement Hash
+    values: Vec<ValidValuesArgument>,
+}
+
+impl ValidValues {
+    pub fn values(&self) -> impl Iterator<Item = &ValidValuesArgument> {
+        self.values.iter()
+    }
+}
+
+impl<V: IslVersion> TypeDefinitionBuilder<V> {
+    /// Adds a `valid_values` constraint to this type definition.
+    pub fn valid_values<T: Into<ValidValuesArgument>, I: IntoIterator<Item = T>>(
+        mut self,
+        values: I,
+    ) -> Self {
+        let values = values.into_iter().map(|it| it.into()).collect();
+        let constraint = ValidValues { values };
+        self.constraints.push(constraint.into());
+        self
+    }
+
+    /// Adds a `valid_values` constraint with a single range argument to this type definition.
+    pub fn valid_number_range<
```

Comment:
🗺️ I think that restricting the range of a number is a common enough use case to justify having a specialized function for it.